---
title: "PPFE_Ligand_Receptor"
author: "Jannik Ruwisch"
date: "2024-12-23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = T,
                      fig.width = 14,
                      fig.height = 7,
                      dev = "png",
                      cache = F)
```

# Load libraries
```{r}
library(multinichenetr)
library(SingleCellExperiment)
library(dplyr)
library(ggplot2)
library(nichenetr)
library(data.table)
library(magrittr)
library(SCP)
library(Seurat)
library(SeuratObject)
library(tidyverse)
library(tidyseurat)
library(ggpubr)
set.seed(123)
```

# Select Folder
```{r}
output_dir <-  choose.dir()
path <- choose.dir()
```


# 1. Read in the Dataset
```{r}
test <- readRDS("./PPFE_merged_final")

# Subset PPFE and CTRL 
test <- subset(test, disease.ident != "IPF")

# Set the Identity and Downsample
Idents(test) <- "orig.ident_final"
df <- data.frame(condition=test$disease.ident, 
                 cell_type=test$Celltype_ID_final_heat, 
                 pat = test$orig.ident_final)
df$barcode <- rownames(df)

subs_df <- df %>% 
  group_by(condition, cell_type, pat) %>% 
  slice_sample(n=100) #  sample for each condition/celltype

# Subset the Seurat object based on the barcodes
test_small <- subset(test, cells = subs_df$barcode)

# Inspect the subsetted Seurat object
DimPlot(test_small, reduction = "ReUMAP_50", split.by = "disease.ident")

# Save downsampled data
saveRDS(test_small, "PPFE_CTRL_ALL_subsampled_celltype_origIdent_run2.rds")
```

```{r}
test_small$orig.ident_final <- as.factor(test_small$orig.ident_final)
test_small$disease.ident <- as.factor(test_small$disease.ident)
test_small$Celltype_ID_final_heat <- as.factor(test_small$Celltype_ID_final_heat)
# Modifiy some characters for used downstream syntax
test_small <- test_small %>%
  mutate(Celltype_ID_final_niche = factor(case_when(Celltype_ID_final_heat == "CD4+_TC_naive" ~ "CD4_TC_naive",
                                             Celltype_ID_final_heat == "CD4+_TC_effector_memory" ~ "CD4_TC_effector",
                                             Celltype_ID_final_heat == "CD4+_TC_TH17" ~ "CD4_TC_effector",
                                             Celltype_ID_final_heat == "CD4+_TC_TH1" ~ "CD4_TC_effector",
                                             Celltype_ID_final_heat == "CD8+_TC_effector_memory" ~ "CD8_TC_effector_memory",
                                             Celltype_ID_final_heat == "NK_Cells_CD16+" ~ "NKCellsCD16_hi",
                                             Celltype_ID_final_heat == "NK_Cells_CD16-" ~ "NKCellsCD16_lo",
                                             Celltype_ID_final_heat == "CD8+_TC_resident_memory" ~ "CD8_TC_resident_memory",
                                             Celltype_ID_final_heat == "CD4+_TC_Cytotoxic" ~ "CD4_TC_Cytotoxic",
                                             Celltype_ID_final_heat == "Schwann_Neu" ~ "Basal",
                                             Celltype_ID_final_heat == "PNEC" ~ "Basal",
                                             Celltype_ID_final_heat == "Langerhans-Cells" ~ "LangerhansCells",
                                             .default = Celltype_ID_final_heat)
                                             )
                                          )
         
# Convert to sce object
sce <- as.SingleCellExperiment(test_small, assay = "RNA")
```

# 2. Setup  
```{r}
organism = "human"

options(timeout = 120)

if(organism == "human"){
  
  lr_network_all = 
    readRDS(url(
      "https://zenodo.org/record/10229222/files/lr_network_human_allInfo_30112033.rds"
      )) %>% 
    mutate(
      ligand = convert_alias_to_symbols(ligand, organism = organism), 
      receptor = convert_alias_to_symbols(receptor, organism = organism))
  
  lr_network_all = lr_network_all  %>% 
    mutate(ligand = make.names(ligand), receptor = make.names(receptor)) 
  
  lr_network = lr_network_all %>% 
    distinct(ligand, receptor)
  
  ligand_target_matrix = readRDS(url(
    "https://zenodo.org/record/7074291/files/ligand_target_matrix_nsga2r_final.rds"
    ))
  
  colnames(ligand_target_matrix) = colnames(ligand_target_matrix) %>% 
    convert_alias_to_symbols(organism = organism) %>% make.names()
  rownames(ligand_target_matrix) = rownames(ligand_target_matrix) %>% 
    convert_alias_to_symbols(organism = organism) %>% make.names()
  
  lr_network = lr_network %>% filter(ligand %in% colnames(ligand_target_matrix))
  ligand_target_matrix = ligand_target_matrix[, lr_network$ligand %>% unique()]
  
} else if(organism == "mouse"){
  
  lr_network_all = readRDS(url(
    "https://zenodo.org/record/10229222/files/lr_network_mouse_allInfo_30112033.rds"
    )) %>% 
    mutate(
      ligand = convert_alias_to_symbols(ligand, organism = organism), 
      receptor = convert_alias_to_symbols(receptor, organism = organism))
  
  lr_network_all = lr_network_all  %>% 
    mutate(ligand = make.names(ligand), receptor = make.names(receptor)) 
  lr_network = lr_network_all %>% 
    distinct(ligand, receptor)
  
  ligand_target_matrix = readRDS(url(
    "https://zenodo.org/record/7074291/files/ligand_target_matrix_nsga2r_final_mouse.rds"
    ))
  
  colnames(ligand_target_matrix) = colnames(ligand_target_matrix) %>% 
    convert_alias_to_symbols(organism = organism) %>% make.names()
  rownames(ligand_target_matrix) = rownames(ligand_target_matrix) %>% 
    convert_alias_to_symbols(organism = organism) %>% make.names()
  
  lr_network = lr_network %>% filter(ligand %in% colnames(ligand_target_matrix))
  ligand_target_matrix = ligand_target_matrix[, lr_network$ligand %>% unique()]
  
}
```

# 3. Multinichenet Pipeline for

## Annotation
```{r}
sample_id = "orig.ident_final"
group_id = "disease.ident"
celltype_id = "Celltype_ID_final_niche" 
covariates = NA 
batches = NA
contrasts_oi = c("'PPFE-CTRL','CTRL-PPFE'")
contrast_tbl = tibble(contrast = c("PPFE-CTRL","CTRL-PPFE"), group = c("PPFE","CTRL"))
```

```{r}
senders_oi =  test_small$Celltype_ID_final_niche %>% unique()  
senders_oi
receivers_oi =test_small$Celltype_ID_final_niche %>% unique()
receivers_oi
sce = sce[, SummarizedExperiment::colData(sce)[,celltype_id] %in% 
            c(senders_oi, receivers_oi)
          ]

conditions_keep = c("PPFE", "CTRL")
sce = sce[, SummarizedExperiment::colData(sce)[,group_id] %in% 
            conditions_keep
          ]


```

## Celltype-Filering
```{r, fig.height=28}
min_cells = 10

abundance_info = get_abundance_info(
  sce = sce, 
  sample_id = sample_id, group_id = group_id, celltype_id = celltype_id, 
  min_cells = min_cells, 
  senders_oi = senders_oi, receivers_oi = receivers_oi, 
  batches = batches
  )
abundance_info$abund_plot_sample
```


## Celltype Filtering based on Abundance
```{r}
abundance_df_summarized = abundance_info$abundance_data %>% 
  mutate(keep = as.logical(keep)) %>% 
  group_by(group_id, celltype_id) %>% 
  summarise(samples_present = sum((keep)))

celltypes_absent_one_condition = abundance_df_summarized %>% 
  filter(samples_present == 0) %>% pull(celltype_id) %>% unique() 
# find truly condition-specific cell types by searching for cell types 
# truely absent in at least one condition

celltypes_present_one_condition = abundance_df_summarized %>% 
  filter(samples_present >= 2) %>% pull(celltype_id) %>% unique() 
# require presence in at least 2 samples of one group so 
# it is really present in at least one condition

condition_specific_celltypes = intersect(
  celltypes_absent_one_condition, 
  celltypes_present_one_condition)

total_nr_conditions = SummarizedExperiment::colData(sce)[,group_id] %>% 
  unique() %>% length() 

absent_celltypes = abundance_df_summarized %>% 
  filter(samples_present < 2) %>% 
  group_by(celltype_id) %>% 
  count() %>% 
  filter(n == total_nr_conditions) %>% 
  pull(celltype_id)
```

```{r}
analyse_condition_specific_celltypes = FALSE

if(analyse_condition_specific_celltypes == TRUE){
  senders_oi = senders_oi %>% setdiff(absent_celltypes)
  receivers_oi = receivers_oi %>% setdiff(absent_celltypes)
} else {
  senders_oi = senders_oi %>% 
    setdiff(union(absent_celltypes, condition_specific_celltypes))
  receivers_oi = receivers_oi %>% 
    setdiff(union(absent_celltypes, condition_specific_celltypes))
}

sce = sce[, SummarizedExperiment::colData(sce)[,celltype_id] %in% 
            c(senders_oi, receivers_oi)
          ]
```

## Gene Filtering
```{r}
min_sample_prop = 0.50 # Among Samples ~50% in the group with lowest samples have to epxress the gene
fraction_cutoff = 0.05 # Within a Sample 5% have to express a gene -> switch to 0.10 if low expressed genes are of interest

frq_list = get_frac_exprs(
  sce = sce, 
  sample_id = sample_id, celltype_id =  celltype_id, group_id = group_id, 
  batches = batches, 
  min_cells = min_cells, 
  fraction_cutoff = fraction_cutoff, min_sample_prop = min_sample_prop)
```

```{r}
genes_oi = frq_list$expressed_df %>% 
  filter(expressed == TRUE) %>% pull(gene) %>% unique() 
sce = sce[genes_oi, ]
```

## Pseudobulk expression calculation gene expression relative to pseudobulk per sample 
```{r}
abundance_expression_info = process_abundance_expression_info(
  sce = sce, 
  sample_id = sample_id, group_id = group_id, celltype_id = celltype_id, 
  min_cells = min_cells, 
  senders_oi = senders_oi, receivers_oi = receivers_oi, 
  lr_network = lr_network, 
  batches = batches, 
  frq_list = frq_list, 
  abundance_info = abundance_info)

# Inspect the results
abundance_expression_info$celltype_info$pb_df %>% head()
abundance_expression_info$sender_receiver_info$pb_df %>% head()
abundance_expression_info$sender_receiver_info$pb_df_group %>% head()
```

## Genome-Wide DEG Analysis
```{r, fig.width=20}
DE_info = get_DE_info(
  sce = sce, 
  sample_id = sample_id, group_id = group_id, celltype_id = celltype_id, 
  batches = batches, covariates = covariates, 
  contrasts_oi = contrasts_oi, 
  min_cells = min_cells, 
  expressed_df = frq_list$expressed_df)

# Check Results
DE_info$celltype_de$de_output_tidy %>% head()
DE_info$hist_pvals
```

## Adjust empirical p-value if needed
```{r}
empirical_pval = FALSE # Check Distribution of p-values -> if not regular >> switch to irregular p-value vignette

if(empirical_pval == TRUE){
  DE_info_emp = get_empirical_pvals(DE_info$celltype_de$de_output_tidy)
  celltype_de = DE_info_emp$de_output_tidy_emp %>% select(-p_val, -p_adj) %>% 
    rename(p_val = p_emp, p_adj = p_adj_emp)
} else {
  celltype_de = DE_info$celltype_de$de_output_tidy
} 
```

## Combine DE-information for Lig-Senders % Rec_recievers
```{r}
sender_receiver_de = multinichenetr::combine_sender_receiver_de(
  sender_de = celltype_de,
  receiver_de = celltype_de,
  senders_oi = senders_oi,
  receivers_oi = receivers_oi,
  lr_network = lr_network
)
```



## Ligand-Receptor Activity Prediction
```{r}
logFC_threshold = 1.35
p_val_threshold = 0.05
p_val_adj = TRUE 

geneset_assessment = contrast_tbl$contrast %>% 
  lapply(
    process_geneset_data, 
    celltype_de, logFC_threshold, p_val_adj, p_val_threshold
  ) %>% 
  bind_rows() 
geneset_assessment
```

## Ligand-Activity Analysis + Ligand-Target Inference
```{r}
top_n_target = 250
verbose = TRUE
cores_system = 16 # This is only recommended if you have many receiver cell types of interest.
n.cores = min(cores_system, celltype_de$cluster_id %>% unique() %>% length()) 

ligand_activities_targets_DEgenes = suppressMessages(suppressWarnings(
  get_ligand_activities_targets_DEgenes(
    receiver_de = celltype_de,
    receivers_oi = intersect(receivers_oi, celltype_de$cluster_id %>% unique()),
    ligand_target_matrix = ligand_target_matrix,
    logFC_threshold = logFC_threshold,
    p_val_threshold = p_val_threshold,
    p_val_adj = p_val_adj,
    top_n_target = top_n_target,
    verbose = verbose, 
    n.cores = n.cores
  )
))
```


## Prioritization: rank cell-cell communication patterns through multi-criteria prioritization
```{r}
ligand_activity_down = FALSE
sender_receiver_tbl = sender_receiver_de %>% distinct(sender, receiver)
metadata_combined = SummarizedExperiment::colData(sce) %>% tibble::as_tibble()

if(!is.na(batches)){
  grouping_tbl = metadata_combined[,c(sample_id, group_id, batches)] %>% 
    tibble::as_tibble() %>% distinct()
  colnames(grouping_tbl) = c("sample","group",batches)
} else {
  grouping_tbl = metadata_combined[,c(sample_id, group_id)] %>% 
    tibble::as_tibble() %>% distinct()
  colnames(grouping_tbl) = c("sample","group")
}

prioritization_tables = suppressMessages(multinichenetr::generate_prioritization_tables(
    sender_receiver_info = abundance_expression_info$sender_receiver_info,
    sender_receiver_de = sender_receiver_de,
    ligand_activities_targets_DEgenes = ligand_activities_targets_DEgenes,
    contrast_tbl = contrast_tbl,
    sender_receiver_tbl = sender_receiver_tbl,
    grouping_tbl = grouping_tbl,
    scenario = "regular", # all prioritization criteria will be weighted equally
    fraction_cutoff = fraction_cutoff, 
    abundance_data_receiver = abundance_expression_info$abundance_data_receiver,
    abundance_data_sender = abundance_expression_info$abundance_data_sender,
    ligand_activity_down = ligand_activity_down
  ))

prioritization_tables$group_prioritization_tbl %>% head(20)
```

## across-Sample Correlation between ligands and receptors
```{r}
lr_target_prior_cor = lr_target_prior_cor_inference(
  receivers_oi = prioritization_tables$group_prioritization_tbl$receiver %>% unique(), 
  abundance_expression_info = abundance_expression_info, 
  celltype_de = celltype_de, 
  grouping_tbl = grouping_tbl, 
  prioritization_tables = prioritization_tables, 
  ligand_target_matrix = ligand_target_matrix, 
  logFC_threshold = logFC_threshold, 
  p_val_threshold = p_val_threshold, 
  p_val_adj = p_val_adj
  )
```

## Save the nichenet ouput
```{r}
multinichenet_output = list(
    celltype_info = abundance_expression_info$celltype_info,
    celltype_de = celltype_de,
    sender_receiver_info = abundance_expression_info$sender_receiver_info,
    sender_receiver_de =  sender_receiver_de,
    ligand_activities_targets_DEgenes = ligand_activities_targets_DEgenes,
    prioritization_tables = prioritization_tables,
    grouping_tbl = grouping_tbl,
    lr_target_prior_cor = lr_target_prior_cor
  ) 
multinichenet_output = make_lite_output(multinichenet_output)

save = TRUE
if(save == TRUE){
  saveRDS(multinichenet_output, paste0(path, "PPFE_CTRL_subsampled_Celltype_multinichenet_output.rds"))

}
```

```{r}
multinichenet_output<- readRDS("PPFE_CTRL_subsampled_Celltype_multinichenet_output.rds")
```